src/server/index.js

```diff
@@ -109,96 +115,219 @@ const createStock = quote => {
   };
 };
 
-const calculateDateYearsInPast = years => {
-  const year = new Date().getFullYear() - years;
-  const date = new Date();
-  date.setFullYear(year);
-  return date;
+// Used for historical() data obtained using period 'd'
+const getDatesAndPrices = dailyData => {
+  let datesAndPrices = [];
+  dailyData.forEach(({
+    date,
+    close
+  }) => {
+    datesAndPrices.unshift({
+      date: formatDateForMaxStockData(date),
+      price: close
+    });
+  });
+  return datesAndPrices;
+};
+
+// NOTE: this date has timezone UTC, which is incorrect but works in this
+// case because we are just extracting the time
+const getAdjustedDateForTimestamp = (gmtoffset, timestamp) => {
+  const adjustedTimestamp =
+    (timestamp + gmtoffset) * constants.MILLISECONDS_PER_SECOND;
+  return new Date(adjustedTimestamp);
 };
 
-const padSingleDigitWithZero = value => {
-  let num = parseInt(value);
-  // need to check value because parseInt turns '12hello' into a number
-  if (isNaN(value) || isNaN(num)) {
-    throw new TypeError(`${padSingleDigitWithZero.name} requires a number or numeric string`);
+const getDateAndTime = (gmtoffset, timestamp, dateAndTimeFormat) => {
+  const dateAndTime = getAdjustedDateForTimestamp(gmtoffset, timestamp);
+  return dateFormat(dateAndTime, dateAndTimeFormat, true);
+};
+
+const getStartOfDayTimestampIndex = (dayIndex, timestampsPerDay) =>
+  Math.floor(dayIndex * timestampsPerDay);
+
+const getEndOfDayTimestampIndex = (dayIndex, timestampsPerDay) =>
+  Math.floor((dayIndex + 1) * timestampsPerDay);
+
+// Could be buggy depending on how the timestamps in their API are given.
+// Couldn't find info on this so will need to test at certain times of day.
+const getDatesAndTimesForOneDay = (
+  close,
+  dayIndex,
+  gmtoffset,
+  numberOfDays,
+  timestamp,
+  timestampIntervals
+) => {
+  let datesTimesAndPrices = [];
+  const timestampsPerDay = timestamp.length / numberOfDays;
+  const dateAndTimeFormat = (numberOfDays === constants.ONE_DAY)
+    ? constants.DATE_FORMAT_ONE_DAY
+    : constants.DATE_FORMAT_FIVE_DAY;
+
+  for (
+    let i = getStartOfDayTimestampIndex(dayIndex, timestampsPerDay);
+    i < getEndOfDayTimestampIndex(dayIndex, timestampsPerDay);
+    i++
+  ) {
+    if (timestamp[i] < timestampIntervals[dayIndex].start
+      || timestamp[i] > timestampIntervals[dayIndex].end) {
+      continue;
+    }
+    datesTimesAndPrices.push({
+      dateAndTime: getDateAndTime(gmtoffset, timestamp[i], dateAndTimeFormat),
+      price: close[i]
+    });
   }
-  return num < 10 ? '0' + num : num.toString();
+  return datesTimesAndPrices;
 };
 
-const formatDate = date => {
-  if (!(date instanceof Date)) {
-    throw new TypeError(`${formatDate.name} requires a date`);
+// days are 0 indexed
+const getEndTimestampForDay = (dayIndex, meta) =>
+  meta.tradingPeriods.regular[dayIndex][0].end;
```

Comment:
what's the `[0]`?

Reply:
The index of the first element of that array? Idk, it's arbitrary and just the way the API is set up, with all the data in that first element of that array. No clue what a descriptive constant would be for the 0, so I left it like this. thoughts?

Reply:
just add a comment :)